### PR TITLE
Improve 1.9 -> 1.8 chunk translation and fixed #2065

### LINKED ISSUE #2065: 1.9+ client does not fully load chunks on 1.8.8 paperspigot/tacospigot server
You told me test with 10 chunks, so I tested with it.

How can we reproduce it?

Join with 1.8 & 1.9+ and compare the chunk loading
This is 1.8.9 Video : https://www.youtube.com/watch?v=Y7VA7LQh1Yo
This is 1.12.2 Video : https://www.youtube.com/watch?v=fHfAf4APqtQ

White Gray wool = 8 chunks
Cyan wool = 9 chunks
Purple wool = 10 chunks

You see that it missing 1 chunks on 1.9+ even serverside view-distance is set to 10 on spigot.yml & server.propertise

Expected behaviour
Should fully load chunks like 1.8.9

ViaVersion Dump
https://dump.viaversion.com/f63d332e1b1aa91de90d8aa1d169044ed849c56a0b6a47a6c892fc4e5c60a9ff

Checklist

[Yes] I have included a Viaversion dump
[Yes] If applicable, I have included a paste (not a screenshot) of the error
[Yes] I have tried a build from https://ci.viaversion.com/ and the issue still persists

### PR DESCRIPTION
**Pull request by RaphiMC**
This pull request contains many changes made to 1.9 -> 1.8 chunk translation:
- Replaced Chunk1_9To1_8 with seperate types to match normal Via codestyle
- Improved the handling of special chunks (unload chunks)
- Add own type for bulk chunks to improve translation speed and allow the use of type caching in lower (ViaRewind) protocols
- Fixed #2065 by sending empty chunks around the newly loaded chunks in order to force the client to render the normal chunks

The changes were tested on many different servers and also on spigot/vanilla (all 3 dimensions) and bungeecord networks.

### CODE REVIEW

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/types/ChunkBulk1_8Type.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        private final int chunkZ;
+        private final int bitmask;
+        private final byte[] data;
+
+        public ChunkBulkSection(final ByteBuf input, final boolean skyLight) {
+            this.chunkX = input.readInt();
+            this.chunkZ = input.readInt();
+            this.bitmask = input.readUnsignedShort();
+            final int bitCount = Integer.bitCount(this.bitmask);
+            this.data = new byte[(bitCount * ((4096 * 2) + 2048)) + (skyLight ? bitCount * 2048 : 0) + 256];
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/types/ChunkBulk1_8Type.java:98]: Make comments explaining the numbers/put them into private static final fields

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/WorldPackets.java [hunk 2]
```diff
@@ -132,22 +137,52 @@ public void registerMap() {
...
-                            // eat any other data (Usually happens with unload packets)
+                            wrapper.write(new Chunk1_9_1_2Type(clientWorld), chunk);
+
+                            clientChunks.getLoadedChunks().add(ClientChunks.toLong(chunk.getX(), chunk.getZ()));
+
+                            // Send empty chunks surrounding the loaded chunk to force 1.9+ clients to render the new chunk
+                            if (Via.getConfig().isChunkBorderFix()) {
+                                for (BlockFace face : BlockFace.values()) {
+                                    if (face.getAxis().equals(BlockFace.EnumAxis.Y)) continue;
+                                    if (!clientChunks.getLoadedChunks().contains(ClientChunks.toLong(chunk.getX() + face.getModX(), chunk.getZ() + face.getModZ()))) {
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/WorldPackets.java:177]: Put the chunk coordinate into a variable for better readability, same in the other places.

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/WorldPackets.java [hunk 3]
```diff
@@ -132,22 +137,52 @@ public void registerMap() {
...
                         } else {
-                            wrapper.write(type, chunk);
-                            // eat any other data (Usually happens with unload packets)
+                            wrapper.write(new Chunk1_9_1_2Type(clientWorld), chunk);
+
+                            clientChunks.getLoadedChunks().add(ClientChunks.toLong(chunk.getX(), chunk.getZ()));
+
+                            // Send empty chunks surrounding the loaded chunk to force 1.9+ clients to render the new chunk
+                            if (Via.getConfig().isChunkBorderFix()) {
+                                for (BlockFace face : BlockFace.values()) {
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/WorldPackets.java:175]: Make a private static final array containing the 4 cardinal directions (best put into the BlockFace enum itself), use that in the other places as well.

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/PlayerPackets.java [hunk 4]
```diff
@@ -300,22 +307,6 @@ public void registerMap() {
             }
         });
 
-        protocol.registerClientbound(ClientboundPackets1_8.UPDATE_HEALTH, new PacketRemapper() {
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/PlayerPackets.java:303]: Is this not/no longer necessary?

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/types/ChunkBulk1_8Type.java (hunk 1 above)
Comment (RaphiMC) [reply common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/types/ChunkBulk1_8Type.java:98]: Those numbers are from the previous chunk reading code, idk what they do exactly

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/PlayerPackets.java (hunk 4 above)
Comment (RaphiMC) [reply common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/PlayerPackets.java:303]: The client doesn't unload chunks when it gets a health = 0 packet, it only shows the death screen, the respawn packet unload the chunks when the client presses respawn

**Review by RaphiMC — COMMENTED** [review]
Made the changes

**Review by kennytv — APPROVED** [review]
Thank you

**Comment by tomcraft** [comment]
Great job ! Is the option required for the improvements or does it enable another "level" of improvement ?

**Comment by RaphiMC** [comment]
The option smoothens chunk loading and fixes #2065

**Comment by tomcraft** [comment]
Ok so if i don't enable the option, does it change anything compared to previous behavior ?

**Comment by RaphiMC** [comment]
No, its only faster

**Comment by tomcraft** [comment]
Ok thank you for the clarification.

**Pull request merged**